### PR TITLE
Added architecture explanation

### PR DESCRIPTION
**Pull request by ishaankolluri**
Describes the architecture of the Extend compiler.
Possible pain points:

* the usage of `linker.ml`. Not sure if I got it entirely correct
* Haven't mentioned `build_graph.ml`. Investigated a bit and wasn't sure how it fit rn into the compilation process - would like some clarification on that front to better write this section.

### CODE REVIEW

File: tex/architecture.tex [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  \item \texttt{scanner.mll}: OCamllex scanner - consumes tokens.
+  \item \texttt{parser.mly}: OCamlyacc parser - represents the Extend grammar.
+  \item \texttt{ast.ml}: Abstract Syntax Tree, created from the output of the parser and representing the structure of an Extend program.
+  \item \texttt{transform.ml}: Analyzes semantics and performs syntactic desugaring for easier compilation.
+  \item \texttt{codegen.ml}: The LLVM IR code generator.
+  \item \texttt{linker.ml}: Calls intermediary compilation steps on the generated \texttt{.ll}, including external functions if needed.
+\end{itemize}
+
+  \subsection{The Scanner}
+  The function of \texttt{scanner.ll} is to parse a text stream into various tokens to be used in an Extend program.
```
Comment (Neitsch) [inline tex/architecture.tex]: scanner.ll? -> scanner.mll

File: tex/architecture.tex [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  Only the tokens that are valid in Extend are to be given to the parser; all others will return a syntax error marked by the line and character number.
+
+  \subsection{The Parser and Abstract Syntax Tree}
+  The parser converts the tokens read by the scanner into a syntax tree deemed acceptable grammar within the Extend Language. This is converted into an Abstract Syntax Tree, which has nodes that can be consumed by the back end of the Extend compiler.
+
+  \subsection{The Transformer}
+  In Extend, the transformer serves multiple purposes. It expands compact statements in the Extend syntax tree into statements with equivalent functionality, but better compilability. This step is done to preserve the convenience for the user, but revert the code later into a form that is easier for the compiler to chew on.
```
Comment (Neitsch) [inline tex/architecture.tex]: equivalent functionality, but reduced breadth.

File: tex/architecture.tex [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  \newline \newline
+  Additionally, the transformer serves as the semantic analyzer. It ensures that Extend functions, variables, expressions, and more are being used properly at compile time, and throws flavorful exceptions to the user so that they may better understand why their program was illegal.
+
+  \subsection{The Code Generator}
+  Provided that the program was deemed legal by \texttt{transform.ml}, the code generator will take the program definition in the abstract syntax tree and generate the appopriate LLVM IR to turn it into a functional program. Instructions to allocate memory, interact with external functions, and platform optimization can be found here.
+
+  \subsection{The Linker}
+  If successful LLVM IR is generated, the linker will adopt the role of creating the \texttt{.ll} to an object file, and compile other resources provided that the program needs it. For example, if external functions are used, the linker will include it as part of the compilation process.
```
Comment (Neitsch) [inline tex/architecture.tex]: adopt the role of building an executable object from the .ll file. This includes compiling it to an object file and linking the runtime environment along with other imported libraries

**Comment by ishaankolluri** [comment]
I need to update this for `semant.ml`, right? The semantic analysis got refactored out of `transform.ml`, if i've read the commit correctly

**Review by Neitsch — APPROVED** [review]

**Pull request merged**